service/src/main/java/org/cbioportal/service/impl/GenericAssayEnrichmentServiceImpl.java

```diff
@@ -0,0 +1,170 @@
+            return enrichmentDatum;
+        }).collect(Collectors.toList());
+
+    }
+
+    @Override
+    @Transactional(readOnly = true)
+    public List<GenericAssayCategoricalEnrichment> getGenericAssayCategoricalEnrichments(String molecularProfileId,
+                                                                                         Map<String, List<MolecularProfileCaseIdentifier>> molecularProfileCaseSets, EnrichmentType enrichmentType)
+        throws MolecularProfileNotFoundException {
+
+        MolecularProfile molecularProfile = getAndValidateMolecularProfile(molecularProfileId, false);
+
+        Iterable<GenericAssayMolecularAlteration> maItr = molecularDataRepository
+            .getGenericAssayMolecularAlterationsIterable(molecularProfile.getStableId(), null, "SUMMARY");
+
+        Map<String, List<MolecularProfileCaseIdentifier>> filteredMolecularProfileCaseSets;
+        filteredMolecularProfileCaseSets = filterMolecularProfileCaseSets(molecularProfile, molecularProfileCaseSets);
+
+        List<GenericAssayCategoricalEnrichment> genericAssayCategoricalEnrichments = expressionEnrichmentUtil.getGenericAssayCategoricalEnrichments(molecularProfile,
+            filteredMolecularProfileCaseSets, enrichmentType, maItr);
+
+        calcQValues(genericAssayCategoricalEnrichments);
+
+        List<String> getGenericAssayStableIds = genericAssayCategoricalEnrichments.stream()
+            .map(GenericAssayEnrichment::getStableId).collect(Collectors.toList());
+        Map<String, GenericAssayMeta> genericAssayMetaByStableId = getGenericAssayMetaByStableId(getGenericAssayStableIds, molecularProfileId);
+
+        return genericAssayCategoricalEnrichments.stream().map(enrichmentDatum -> {
+            enrichmentDatum.setGenericEntityMetaProperties(
+                genericAssayMetaByStableId.get(enrichmentDatum.getStableId()).getGenericEntityMetaProperties());
+            return enrichmentDatum;
+        }).collect(Collectors.toList());
+    }
+
+    private MolecularProfile getAndValidateMolecularProfile(String molecularProfileId, boolean isBinary) throws MolecularProfileNotFoundException {
+        MolecularProfile molecularProfile = molecularProfileService.getMolecularProfile(molecularProfileId);
+        validateMolecularProfile(molecularProfile, Arrays.asList(MolecularProfile.MolecularAlterationType.GENERIC_ASSAY), isBinary);
+        return molecularProfile;
+    }
+
+    private void validateMolecularProfile(MolecularProfile molecularProfile,
+                                          List<MolecularProfile.MolecularAlterationType> validMolecularAlterationTypes, 
+                                          boolean isBinary) throws MolecularProfileNotFoundException {
+        if (!validMolecularAlterationTypes.contains(molecularProfile.getMolecularAlterationType())) {
+            throw new MolecularProfileNotFoundException(molecularProfile.getStableId());
+        }
+        if(isBinary) {
```

Comment:
There are two types of generic assay data calling this validation, so the second one is used to verify all the data fetched by getGenericAssayBinaryEnrichments are Binary type, not Categorical, Limited Value or null. This will ensure frontend display the right things.